wandb/summary.py

```diff
@@ -14,199 +16,295 @@
+        distinguishable from that path being set to the value None.
+        """
+        raise NotImplementedError
+
+    def _root_set(self, path, new_keys_values):
+        """Set a value at a particular path in the root.
+
+        This should only be implemented by the "_root" child class.
+        """
+        raise NotImplementedError
+
+    def _root_del(self, path):
+        """Delete a value at a particular path in the root.
+
+        This should only be implemented by the "_root" child class.
+        """
+        raise NotImplementedError
 
     def _write(self, commit=False):
+        # should only be implemented on the root summary
         raise NotImplementedError
 
-    def _transform(self, k, v=None, write=True):
-        """Transforms keys json into rich objects for the data api"""
-        if not write and isinstance(v, dict):
-            if v.get("_type") in H5_TYPES:
-                return self.read_h5(k, v)
-            elif v.get("_type") == 'parquet':
-                print(
-                    'This dataframe was saved via the wandb data API. Contact support@wandb.com for help.')
-                return None
-            # TODO: transform wandb objects and plots
-            else:
-                return {key: self._transform(k + "." + key, value, write=False) for (key, value) in v.items()}
+    def keys(self):
+        return self._dict.keys()
 
-        return v
+    def get(self, k, default=None):
+        k = k.strip()
+        if k not in self._dict:
+            self._root._root_get(self._path + (k,), self._dict)
+        return self._dict.get(k, default)
 
     def __getitem__(self, k):
-        return self._transform(k, self._summary[k], write=False)
+        self.get(k)  # load the value into _dict if it should be there
+        return self._dict[k]
+
+    def __contains__(self, k):
+        k = k.strip()
+        return k in self._dict
 
     def __setitem__(self, k, v):
-        key = k.strip()
-        self._summary[key] = self._transform(key, v)
-        self._locked_keys.add(key)
-        self._write()
+        k = k.strip()
+        path = self._path
 
-    def __setattr__(self, k, v):
-        if k.startswith("_"):
-            super(Summary, self).__setattr__(k, v)
+        if isinstance(v, dict):
+            self._dict[k] = SummarySubDict(self._root, path)
+            self._root._root_set(path, [(k, {})])
+            self._dict[k].update(v)
         else:
-            key = k.strip()
-            self._summary[key] = self._transform(key, v)
-            self._locked_keys.add(key)
-            self._write()
+            self._dict[k] = v
+            self._root._root_set(path, [(k, v)])
 
-    def __getattr__(self, k):
-        if k.startswith("_"):
-            return super(Summary, self).__getattr__(k)
-        else:
-            return self._transform(k.strip(), self._summary[k.strip()], write=False)
+        self._locked_keys.add(k)
+
+        self._root._write()
+
+        return v
```

Comment:
is that return needed or doing anything?  normally it isnt set with __setitem__ definitions.